src/machine/asm/mod.rs

```diff
@@ -76,36 +76,39 @@ impl Memory<u64> for Box<AsmCoreMachine> {
             if self.flags[page] & FLAG_FREEZED != 0 {
                 return Err(Error::InvalidPermission);
             }
+        }
+        fill_page_data(self, addr, size, source, offset_from_addr)?;
+        for page_addr in (addr..addr + size).step_by(RISCV_PAGESIZE) {
```

Comment:
```suggestion
for page in (addr..(addr + size - 1) / page_size + 1) {
  self.flags[page] = flags;
}
```

Reply:
Comment added explaining the code, the suggested change is actually slower in a Rust setup.